File: Es4.cpp
/*#include <iostream>
#include "TGraph.h"
#include "TH1D.h"
#include "TF1.h"
#include "TApplication.h"
#include "TMinuit.h"
#include <fstream>
#include <vector>
using namespace std;*/
{

namespace data{
  vector<double> x;
}


void logl(int &npar, double *gin, double &f, double *par, int iflag){
  
  f = 0.0;
  for(int i=0;i<data::x.size();i++){
    f+=3*log(par[0])+data::x[i]/par[0];
  }
}


//int main(){

    auto app=new TApplication("app",0,NULL);

    auto h=new TH1D("h","h",100,0.,0.);

    ifstream file("Dati_es4.dat");
    double val;
    //vector<double> x;
    
    while(file >> val){
      h->Fill(val);
      data:: x.push_back(val);
      
    }

    h->Draw();
    
    
    TMinuit minuit(1);
    minuit.SetFCN(logl);
    minuit.SetErrorDef(0.5);
    minuit.DefineParameter(0,"theta",(1.0/3.0)*h->GetMean(),(1.0/3.0)*h->GetRMS(),0.,0.);

    minuit.Command("MIGRAD");
    double theta,etheta;
    minuit.GetParameter(0,theta,etheta);
    cout << "theta= " << theta << endl;
    cout << "etheta= "  << etheta << endl;
    

    app->Run(true);
    // return 0;
//}
}

File: Es4.py
from   ROOT    import *
from   iminuit import Minuit
import numpy   as     np
from   math    import *

#faccio un fit di ML

def logl(theta):
    f=0.0
    for i in range(len(x)):
        f=f+3*np.log(theta)+x[i]/theta
    return f

hist=TH1D("hist"," ",100,0.,0.)
x=[]

for line in open("Dati_es4.dat"):
    val=float(line)
    hist.Fill(val)
    x.append(val)

m=Minuit(logl,theta=hist.GetMean()/3.0)
m.errordef=0.5
m.print_level=2
m.migrad()
theta=m.values[0]
etheta=m.errors[0]
fun=TF1("fun","[0]*( (pow(x,2)*exp(-x/[1]))/(2*pow([1],3)))",0,30)
fun.FixParameter(0,hist.GetEntries()*hist.GetBinWidth(1))
fun.FixParameter(1,theta)
hist.Draw()
fun.Draw("same")


print("theta= ",round(theta,1))
print("etheta= ",round(etheta,1))



gApplication.Run(True)




        
